contracts/FIFOMarket.sol

```diff
@@ -96,55 +100,62 @@ contract FIFOMarket is
-    address recipient = abi.decode(userData, (address));
-    require(recipient == address(recipient), "FIFOMarket: Invalid address");
-    require(
-      recipient != address(0),
-      "FIFOMarket: cannot mint to the 0 address"
-    );
-    require(
-      msg.sender == address(_nori),
-      "FIFOMarket: This contract can only receive NORI"
-    ); // todo verify this can only be invoked by the nori contract
     uint256 certificateAmount = (amount * 100) / (100 + _noriFee);
     uint256 remainingAmountToFill = certificateAmount;
-    uint256[] memory ids = new uint256[](_queue.length());
-    uint256[] memory amounts = new uint256[](_queue.length());
-    address[] memory suppliers = new address[](_queue.length());
-    for (uint256 i = 0; i < _queue.length(); i++) {
-      uint256 removalAmount = _removal.balanceOf(address(this), _queue.at(i));
-      address supplier = _removal.vintage(_queue.at(i)).supplier;
+
+    address recipient = abi.decode(userData, (address)); // todo handle the case where someone invokes this function without operatorData
+    require(recipient == address(recipient),"FIFOMarket: Invalid address");
+    require(recipient != address(0), "FIFOMarket: Cannot mint to the 0 address");
+    // todo verify this can only be invoked by the nori contract
+    require(msg.sender == address(_nori), "FIFOMarket: This contract can only receive NORI");
+    uint256[] memory ids = new uint256[](_queueLength());
+    uint256[] memory amounts = new uint256[](_queueLength());
+    address[] memory suppliers = new address[](_queueLength());
+    for (uint256 i = _queueHeadIndex; i < _queueNextInsertIndex; i++) {
+      uint256 removalAmount = _removal.balanceOf(
+        address(this),
+        _queue[i]
+      );
+      address supplier = _removal.vintage(_queue[i]).supplier;
       if (remainingAmountToFill < removalAmount) {
-        ids[i] = _queue.at(i);
+        ids[i] = _queue[i];
         amounts[i] = remainingAmountToFill;
         suppliers[i] = supplier;
         remainingAmountToFill = 0;
-      } else if (remainingAmountToFill >= removalAmount) {
-        if (i == _queue.length() - 1 && remainingAmountToFill > removalAmount) {
+      } else {
+        if (
+            i == _queueNextInsertIndex - 1 &&
+            remainingAmountToFill > removalAmount
+        ) {
           revert("FIFOMarket: Not enough supply");
         }
-        ids[i] = _queue.at(i);
+        ids[i] = _queue[i];
         amounts[i] = removalAmount;
         suppliers[i] = supplier;
         remainingAmountToFill -= removalAmount;
-      } else {
-        revert("FIFOMarket: Not enough supply");
       }
+
       if (remainingAmountToFill == 0) {
         break;
       }
     }
+    // In a completely empty queue, the for loop never runs.
+    // Catch this case here without the gas-heavy numberOfNrtsInQueue call.
+    require(amounts[0] > 0, "FIFOMarket: Not enough supply");
```

Comment:
might be better to check this before the for loop?

Reply:
actually, with this different implementation of the queue I think I can change the way this check is made. the queue is empty if the head index and the next insert index are the same.  I'll add this check before the loop, I think it's probably a tiny bit gas cheaper that way and clearer